ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -159,6 +163,10 @@ public int compare(IndexEntry e1, IndexEntry e2) {
 
   private Runnable getBlobReadInfoTestCallback = null;
 
+  // The lock to protect cachedKeys in the findMissingKeysInBatch
+  private final Lock missingKeysLock = new ReentrantLock();
+  private LinkedList<Pair<Offset, Set<StoreKey>>> cachedKeys = new LinkedList<>();
```

Comment:
Like we were discussing offline, may be we might need separate cached key set for each remote replica since each of them might be at different replication sync offset.

Reply:
Updated.